File: split/makeaccent.pas
procedure makeaccent;
var
  s, t: real;
  p, q, r: halfword;
  f: internalfontnumber;
  a, h, x, w, delta: scaled;
  i: fourquarters;
begin
  scancharnum;
  f := eqtb[3934].hh.rh;
  p := newcharacter(f, curval);
  if p <> 0 then
  begin
    x := fontinfo[5 + parambase[f]].int;
    s := fontinfo[1 + parambase[f]].int / 65536.0;
    a := fontinfo[widthbase[f] + fontinfo[charbase[f] + mem[p].hh.b1].qqqq.b0].int;
    doassignments;
    {1124:}
q := 0;
    f := eqtb[3934].hh.rh;
    if (curcmd = 11) or (curcmd = 12) or (curcmd = 68) then
      q := newcharacter(f, curchr)
    else if curcmd = 16 then
    begin
      scancharnum;
      q := newcharacter(f, curval);
    end
    else
      backinput{:1124}
;
    if q <> 0 then{1125:}

    begin
      t := fontinfo[1 + parambase[f]].int / 65536.0;
      i := fontinfo[charbase[f] + mem[q].hh.b1].qqqq;
      w := fontinfo[widthbase[f] + i.b0].int;
      h := fontinfo[heightbase[f] + (i.b1 - 0) div 16].int;
      if h <> x then
      begin
        p := hpack(p, 0, 1);
        mem[p + 4].int := x - h;
      end;
      delta := round((w - a) / 2.0 + h * t - x * s);
      r := newkern(delta);
      mem[r].hh.b1 := 2;
      mem[curlist.tailfield].hh.rh := r;
      mem[r].hh.rh := p;
      curlist.tailfield := newkern(-a - delta);
      mem[curlist.tailfield].hh.b1 := 2;
      mem[p].hh.rh := curlist.tailfield;
      p := q;
    end{:1125}
;
    mem[curlist.tailfield].hh.rh := p;
    curlist.tailfield := p;
    curlist.auxfield.hh.lh := 1000;
  end;
end;

File: cpp/makeaccent.cpp
#include "makeaccent.h"
#include "scancharnum.h"
#include "newcharacter.h"
#include "doassignments.h"
#include "backinput.h"
#include "hpack.h"
#include "newkern.h"
#include <cmath>

void makeaccent(void)
{
	float s, t;
	halfword r;
	scaled a, h, x, w, delta;
	fourquarters i;
	scancharnum();
	internalfontnumber f = cur_font();
	auto p = newcharacter(f, curval);
	if (p)
	{
		x = param(x_height_code, f);
		s = param(slant_code, f)/float(unity);
		a = char_width(f, char_info(f, character(p)));
		doassignments();
		halfword q = 0;
		f = cur_font();
		if (curcmd == letter || curcmd == other_char || curcmd == char_given)
			q = newcharacter(f, curchr);
		else 
			if (curcmd == char_num)
			{
				scancharnum();
				q = newcharacter(f, curval);
			}
			else
				backinput();
		if (q)
		{
			t = param(slant_code, f)/float(unity);
			i = char_info(f, character(q));
			w = char_width(f, i);
			h = char_height(f, i);
			if (h != x)
			{
				p = hpack(p, 0, 1);
				shift_amount(p) = x-h;
			}
			delta = round((w-a)/2.0+h*t-x*s);
			r = newkern(delta);
			subtype(r) = 2;
			link(tail) = r;
			link(r) = p;
			tail = newkern(-a-delta);
			subtype(tail) = 2;
			link(p) = tail;
			p = q;
		}
		tail_append(p);
		space_factor = 1000;
	}
}
